index.d.ts

```diff
@@ -0,0 +1,52 @@
+type Replacer = (key: string, value: any) => void;
+
+interface Options {
+	/**
+	 * Indentation as a string or number of spaces. Pass in null for no formatting.
+	 */
+	indent?: string | number | null;
+	/**
+	 * Detect indentation automatically if the file exists.
+	 */
+	detectIndent?: boolean;
+	/**
+	 * Sort the keys recursively. Optionally pass in a compare function.
+	 */
+	sortKeys?: boolean | ((a: string, b: string) => number);
+	/**
+	 * Passed into JSON.stringify.
+	 */
+	replacer?: Replacer | Array<number | string>;
+	/**
+	 * Mode used when writing the file.
+	 */
+	mode?: number;
+}
+
+/**
+ * Stringify and write JSON to a file atomically.
+ *
+ * Creates directories for you as needed.
+ *
+ * @example
+ * import writeJsonFile from 'write-json-file';
+ *
+ * writeJsonFile.sync('foo.json', {foo: true});
+ * console.log('done');
+ */
+export function sync(filepath: string, data: any, options?: Options): void;
```

Comment:
`data` can't _really_ be `any`. For example, passing in `undefined` will throw a `TypeError`. We could attempt to express valid JSON values to be more specific here.

Reply:
Consider the fact even in TypeScripts "official" `lib.d.ts` file, the signature for `JSON.stringify( ... )` is

```typescript
JSON.stringify(value: any, replacer?: (key: string, value: any) => any, ...): string;
```
That means even the TypeScript team couldn't come with a reliabl, sound way to describe JSON.

Reply:
There's always the `object` type, however I kinda like using `any` for this param.

Reply:
@brandon93s was hoping to prevent passing `undefined` (which causes `TypeError`), the problem is that no signature of any kind (as of `ts@3.0.1`) can prevent that from happening.

```typescript
function foo(a: object | number | string): void { }

// ts won't complain i.e no red squiggly lines
foo(undefined)
// no complains here either
foo(null)
```

Reply:
Yeah, let's make it strict. Can be a type called `JSONStringifyable` maybe.

Reply:
👍 for making it more strict in the `data` param.

There's two options:

1. We create a type called `JSONStringifyable` (the type is slightly "hidden" when using intellisense)

<img width="740" alt="screen shot 2018-08-21 at 9 57 48 am" src="https://user-images.githubusercontent.com/550895/44406033-afb5c400-a528-11e8-8c32-f83fc51142b3.png">

2. We add the types directly in the `data` param, that way it's clear that the input types are `string`|`object`|`number` (my preferred choice).

<img width="757" alt="screen shot 2018-08-21 at 9 59 41 am" src="https://user-images.githubusercontent.com/550895/44406162-f1466f00-a528-11e8-92c9-d7cbef82d884.png">